RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Service/MatrixSDK/MXTemplateUserProfileService.swift

```diff
@@ -0,0 +1,83 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import Combine
+
+@available(iOS 14.0, *)
+class MXTemplateUserProfileService: TemplateUserProfileServiceProtocol {
+    
+    // MARK: - Properties
+    
+    // MARK: Private
+    
+    private let session: MXSession
+    private var listenerReference: Any!
+    @Published private var presence: TemplateUserProfilePresence = .offline
+    
+    // MARK: Public
+    
+    var userId: String {
+        return session.myUser.userId
+    }
+    
+    var displayName: String? {
+        session.myUser.displayname
+    }
+    
+    var avatarUrl: String? {
+        session.myUser.avatarUrl
+    }
+    
+    var presencePublisher: AnyPublisher<TemplateUserProfilePresence, Never> {
+        $presence.eraseToAnyPublisher()
+    }
+    
+    // MARK: - Setup
+    
+    init(session: MXSession) {
+        self.session = session
+        
+        let listenerReference = session.myUser.listen { [weak self] event in
```

Comment:
Maybe we can use a lazy var here

Reply:
Not quite sure how you mean?

Reply:
```
private lazy listenerReference: Any = {
buildPresenceListener()
}()
```

Reply:
And then where do we call the new property `listenerReference`?

Reply:
And where would this be called? We'd need to do something this to setup the presence updates in the `init`:
```swift
init() {
  let _ = listenerReference
}
```

Which seems to a bit non-obvious that there are side effects.
I should probably actually handle the nil case in case SDK does not return a listener and log an error if setup fails.
I'll move it out of the init into a function that sets up the listener.
Maybe have a look and let me know if you have any more thoughts.